readme: rewrite and update the workflows (#41)
NilFoundation/zkllvm-template: README.md

  - [Step 4: Make an account on the Proof Market](#step-4-make-an-account-on-the-proof-market)
  - [Step 5: Publish the circuit statement](#step-5-publish-the-circuit-statement)
  - [Step 6: Check the information about your statement](#step-6-check-the-information-about-your-statement)
- [Part 2. Application developer workflow](#part-2-application-developer-workflow)
  - [Step 1: See the statements available on the Proof Market](#step-1-see-the-statements-available-on-the-proof-market)
  - [Step 2: Post a proof request](#step-2-post-a-proof-request)
  - [Step 3: Check if the proof is ready](#step-3-check-if-the-proof-is-ready)
  - [Step 4: Download the proof](#step-4-download-the-proof)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Introduction

You will run each step of this tutorial as a command, conveniently wrapped in the `scripts/run.sh` script.
We recommend using it when you go through the tutorial for the first time.

Once you've completed the tutorial, you can repeat it by running all commands manually in the console.
Look at the `🧰 [manual mode]` instructions in collapsed blocks.

> They have all the steps with detailed explanation of commands and their parameters,
> file formats and other things.

Reviewer: ```suggestion
They have all the steps with detailed explanations of commands and their parameters,
file formats, and other things.
```

Author: fixed

## 2. Get the Docker images with `=nil;` toolchain

> In the tutorial we will use Docker images with parts of the `=nil;` toolchain.

Reviewer: ```suggestion
In the tutorial, we will use Docker images with parts of the `=nil;` toolchain.
```

Author: fixed

We recommend using them because they're tested for compatibility,
and they save you time on installing and compiling everything:

> * The `nilfoundation/zkllvm-template` image has the zkLLVM part of the toolchain:

Reviewer: That's the second colon here (first on line 68), maybe replace it with a dash or comma?

Author: fixed

  the zkLLVM compiler (`clang`), `assigner`, and `tranpiler` binaries.

> 

Reviewer: No need for two blank lines here

> 

Reviewer: ```suggestion
```
No need for two blank lines here

Author: same

* The `nilfoundation/proof-market-toolchain` image has all you need to make an account on 
the `=nil;` Proof Market, put your circuit on it, and order a proof.

Both images are versioned according to the products they contain.
In the tutorial, we'll use the latest compatible versions of both images:

```bash
ZKLLVM_VERSION=0.0.58
docker pull ghcr.io/nilfoundation/zkllvm-template:${ZKLLVM_VERSION}

TOOLCHAIN_VERSION=0.0.33
docker pull ghcr.io/nilfoundation/proof-market-toolchain:${TOOLCHAIN_VERSION}
```

# Part 1. Circuit development workflow

In the first part of this tutorial, we'll walk through the development workflow

> of a circuit developer. Most operations will be done on a local machine,
> without using the Proof Market.

Reviewer: ```suggestion
of a circuit developer.
Most operations will be done on a local machine, without using the Proof Market.
```

Author: fixed

We will build a circuit, pack it into a circuit statement,
and then use it to build a proof for a particular input.
Last thing, we'll post the statement on the Proof Market,

> so that zkApp developers will be able to request proofs with this statement.

Reviewer: ```suggestion
so that zk applications developers will be able to request proofs with this statement.
```

Code in `./src` is an example of BLS12-381 signature verification via zkLLVM using

> [Crypto3 C++ cryptography suite](https://github.com/nilfoundation/crypto3) as an SDK.

Reviewer: ```suggestion
[Crypto3 C++ cryptography suite](https://github.com/nilfoundation/crypto3) SDK.
```

> 

Author: removed

## Step 3: Produce and verify a proof locally

Now we have everything ready to produce our first proof.

> As a circuit developer, we want to first build it locally,
> to check that our circuit is working.

Reviewer: ```suggestion
As a circuit developer, we want to first build it locally, to check that our circuit is working.
```
This doesn't seem too long a line

Reviewer: Ok, what's our limit?

We'll use the `proof-generator` CLI, which is a part of the Proof Market toolchain.

```bash
scripts/run.sh --docker prove
```

<details>

>   <summary><code>🧰 [manual mode]</code></summary>

Reviewer: No manual mode here?

Author: lines 250 to 266

Now it's time to work with the `=nil;` Proof Market.

## Step 4: Make an account on the Proof Market

To publish statements and order proofs from the Proof Market, you need an account.
We'll use the command line tools to make an account.

First, run a Docker container with the Proof Market toolchain:

```console
$ cd /opt/zkllvm-template
$ scripts/run.sh run_proof_market_toolchain
```

Great, now you're in the container's console.
Time to make an account:

```console
# cd /proof-market-toolchain
# python3 scripts/signup.py user \
    -u <username> \
    -p <password> \
    -e <email>
```

It should return something like this:

> ```

Reviewer: ```suggestion
```json
```

Author: fixed

{"user":"zkdev","active":true,"extra":{},"error":false,"code":201}
```

This command will save your username and password in two files in the container:
* `/proof-market-toolchain/scripts/.user`
* `/proof-market-toolchain/scripts/.secret`

These files in the container are mounted to `.config/.user` and `.config/.secret` on your machine.
This way, when you stop the container, the files will persist until you run it again.

## Step 5: Publish the circuit statement

Remember the statement that we've packed in step 2?
Let's publish it on the Proof Market.

> ```bash
> python3 scripts/statement_tools.py push \

Reviewer: ```suggestion
```console
$ python3 scripts/statement_tools.py push \
```

Author: right, fixed

Reviewer: Ok

    --file /opt/zkllvm-template/build/template.json
```

> ```

Reviewer: ```suggestion
```

Author: added some descriptive text here

Statement from /opt/zkllvm-template/build/template.json was pushed with id 12345678
```

## Step 6: Check the information about your statement

Let's see how the statement is published:

```bash
python3 scripts/statement_tools.py get \
    --key 12345678
```

> Congratulations! You've built a zk-circuit and published it on the Proof Market.

Reviewer: ```suggestion
Congratulations! You've built a zk circuit and published it on the Proof Market.
```

Author: fixed

Now it's time to have a look at how developers of zero-knowledge applications 
use the Proof Market.

# Part 2. Application developer workflow

In this part, we will act as a developer of a zk application.
Our task is to order a proof on the Proof Market:

1. Find a circuit statement.
   We will be using one that has active proof producers,
   who will respond to our request.
2. Post a request for a proof with given statement and particular input.
3. Check that a request was matched and the proof is ready.
4. Download the proof.

All commands in this section run in the container `nilfoundation/proof-market-toolchain`:

```console
$ cd /opt/zkllvm-template
$ scripts/run.sh run_proof_market_toolchain
```

## Step 1: See the statements available on the Proof Market

First, let's see what statements are available on the Proof Market.

> ```

Reviewer: ```suggestion
```bash
```

python3 scripts/statement_tools.py get
```

## Step 2: Post a proof request

> ```

Reviewer: ```suggestion
```console
```

This command will save your username and password in two files in the container:
* `/proof-market-toolchain/scripts/.user`
* `/proof-market-toolchain/scripts/.secret`

These files in the container are mounted to `.config/.user` and `.config/.secret` on your machine.
This way, when you stop the container, the files will persist until you run it again.

## Step 5: Publish the circuit statement

Remember the statement that we've packed in step 2?
Let's publish it on the Proof Market.

```bash
python3 scripts/statement_tools.py push \
    --file /opt/zkllvm-template/build/template.json
```

```
Statement from /opt/zkllvm-template/build/template.json was pushed with id 12345678
```

## Step 6: Check the information about your statement

Let's see how the statement is published:

```bash
python3 scripts/statement_tools.py get \
    --key 12345678
```

Congratulations! You've built a zk-circuit and published it on the Proof Market.
Now it's time to have a look at how developers of zero-knowledge applications 
use the Proof Market.

# Part 2. Application developer workflow

In this part, we will act as a developer of a zk application.
Our task is to order a proof on the Proof Market:

> 1. Find a circuit statement. We will be using one that has active proof producers,
>    who will respond to our request.

Reviewer: ```suggestion
1. Find a circuit statement.
   We will be using one that has active proof producers, who will respond to our request.
```

Author: fixed

```console
$ cd /opt/zkllvm-template
$ scripts/run.sh run_proof_market_toolchain
```

## Step 1: See the statements available on the Proof Market

First, let's see what statements are available on the Proof Market.
```
python3 scripts/statement_tools.py get
```

## Step 2: Post a proof request

```
$ python3 scripts/request_tools.py push \
    --key 12345678 \
    --cost 10 \
    --file /opt/zkllvm-template/src/main-input.json

Limit request:	 {
    "_key": "99887766",
    "statement_key": "12345678",
    "cost": 10,
    "sender": "zkdev",
    "status": "created"
}
```

## Step 3: Check if the proof is ready

You can check the request status at any time:
```console
$ python3 scripts/request_tools.py get --key 99887766

Limit request:	 {
    "_key": "99887766",
    "statement_key": "12345678",
    "cost": 10,
    "sender": "zkdev",
    "status": "created"
}
```

## Step 4: Download the proof

When the proof is ready, download it:

```console
$ python3 scripts/proof_tools.py get \ 
--request-key 99887766 \
--file /tmp/example.proof
$ ls -l /tmp/example.proof
```

Now the proof can be verified, both off-chain and on-chain.

> Next steps will be added soon.

Reviewer: ```suggestion
Next steps will be added soon.

```

Author: fixed
